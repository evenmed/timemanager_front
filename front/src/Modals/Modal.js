import React, { useState } from "react";
import PropTypes from "prop-types";

const Modal = (props) => {
  const [active, setActive] = useState(false);

  const showModal = () => {
    document.body.classList.add("modal-open");
    setActive(true);
  };

  const hideModal = () => {
    document.body.classList.remove("modal-open");
    setActive(false);
  };

  return (
    <>
      <button
        className={`btn btn-${
          props.toggleButtonType ? props.toggleButtonType : "primary"
        }`}
        onClick={showModal}
      >
        {props.toggleButtonContent}
      </button>
      <div
        className={`modal fade ${active ? "show" : ""}`}
        style={{
          display: "block",
          pointerEvents: active ? "all" : "none",
        }}
        tabIndex="-1"
      >
        <div className="modal-dialog">
          <div className="modal-content">
            <div className="modal-header">
              <h5 className="modal-title">{props.title}</h5>
              <button
                type="button"
                className="close"
                data-dismiss="modal"
                aria-label="Close"
                onClick={hideModal}
              >
                <span aria-hidden="true">&times;</span>
              </button>
            </div>
            {props.children(hideModal)}
          </div>
        </div>
      </div>
    </>
  );
};

const ModalBody = ({ children }) => (
  <div className="modal-body">{children}</div>
);

const ModalFooter = ({ children }) => (
  <div className="modal-footer">{children}</div>
);

Modal.propTypes = {
  toggleButtonContent: PropTypes.element.isRequired,
  toggleButtonType: PropTypes.string,
  title: PropTypes.string.isRequired,
  children: PropTypes.func.isRequired,
};

ModalBody.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
};
ModalFooter.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
};

export default Modal;
export { ModalBody, ModalFooter };
