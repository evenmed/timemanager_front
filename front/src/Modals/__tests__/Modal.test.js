import { shallow, mount } from "enzyme";
import toJSON from "enzyme-to-json";
import Modal, { ModalBody, ModalFooter } from "../Modal";

describe("<Modal/>", () => {
  it("renders with proper class", () => {
    const hideModal = jest.fn();
    const wrapper = shallow(
      <Modal title="Test title" active={false} hideModal={hideModal}>
        <p>Child</p>
      </Modal>
    );
    expect(toJSON(wrapper)).toMatchSnapshot(); // Shouldn't have "show" class
    wrapper.setProps({ active: true });
    expect(toJSON(wrapper)).toMatchSnapshot(); // Should have "show" class
  });
  it("calls hideModal function when necessary", () => {
    const hideModal = jest.fn();
    const wrapper = mount(
      <Modal title="Test title" active={true} hideModal={hideModal}>
        <p>Child</p>
      </Modal>
    );

    // Test that it's called with close btn
    const closeBtn = wrapper.find("button.close");
    expect(closeBtn).toHaveLength(1);
    closeBtn.simulate("click");
    expect(hideModal).toHaveBeenCalledTimes(1);

    // Test that it's called when bg is clicked
    const modalDiv = wrapper.find(".modal");
    expect(modalDiv).toHaveLength(1);
    modalDiv.simulate("click");
    expect(hideModal).toHaveBeenCalledTimes(2);

    // Make sure it isn't called when children are clicked
    const modalDialog = wrapper.find(".modal-dialog");
    expect(modalDialog).toHaveLength(1);
    modalDialog.simulate("click");
    expect(hideModal).toHaveBeenCalledTimes(2);
  });
});

describe("<ModalBody/>", () => {
  it("renders properly", () => {
    const wrapper = shallow(
      <ModalBody>
        <p>Some nice content</p>
      </ModalBody>
    );

    expect(wrapper.text()).toContain("Some nice content");
    const bodyTag = wrapper.find(".modal-body");
    expect(bodyTag).toHaveLength(1);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});

describe("<ModalFooter/>", () => {
  it("renders properly", () => {
    const wrapper = shallow(
      <ModalFooter>
        <p>Some nice footer content</p>
      </ModalFooter>
    );

    expect(wrapper.text()).toContain("Some nice footer content");
    const footerTag = wrapper.find(".modal-footer");
    expect(footerTag).toHaveLength(1);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
