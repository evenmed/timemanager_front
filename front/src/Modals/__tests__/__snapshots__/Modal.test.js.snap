// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal/> renders with proper class 1`] = `
<Fragment>
  <div
    className="modal fade "
    onClick={[Function]}
    style={
      Object {
        "display": "block",
        "pointerEvents": "none",
      }
    }
    tabIndex="-1"
  >
    <div
      className="modal-dialog"
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header"
        >
          <h5
            className="modal-title"
          >
            Test title
          </h5>
          <button
            aria-label="Close"
            className="close"
            data-dismiss="modal"
            onClick={[MockFunction]}
            type="button"
          >
            <span
              aria-hidden="true"
            >
              ×
            </span>
          </button>
        </div>
        <p>
          Child
        </p>
      </div>
    </div>
  </div>
</Fragment>
`;

exports[`<Modal/> renders with proper class 2`] = `
<Fragment>
  <div
    className="modal fade show"
    onClick={[Function]}
    style={
      Object {
        "display": "block",
        "pointerEvents": "all",
      }
    }
    tabIndex="-1"
  >
    <div
      className="modal-dialog"
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header"
        >
          <h5
            className="modal-title"
          >
            Test title
          </h5>
          <button
            aria-label="Close"
            className="close"
            data-dismiss="modal"
            onClick={[MockFunction]}
            type="button"
          >
            <span
              aria-hidden="true"
            >
              ×
            </span>
          </button>
        </div>
        <p>
          Child
        </p>
      </div>
    </div>
  </div>
</Fragment>
`;

exports[`<ModalBody/> renders properly 1`] = `
<div
  className="modal-body"
>
  <p>
    Some nice content
  </p>
</div>
`;

exports[`<ModalFooter/> renders properly 1`] = `
<div
  className="modal-footer"
>
  <p>
    Some nice footer content
  </p>
</div>
`;
